feat(sql lab): Save Dataset Modal Autocomplete should display list when overwritting

Antonio-RiveroMartnez opened the pull request:
### SUMMARY
- Use our Select component as substitute of the Autocomplete one so options are loaded without the user having to trigger a search and also we are more consistent with the rest of the app
- Changing `datasetId` to lowercase so when custom props get into the DOM we don't get warning related to invalid formatting
- We extracted the dropdown out of the radio because it causes invalid click handling when an option is selected
- Add getPopupContainer as prop for Select component

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Before:
![error](https://user-images.githubusercontent.com/38889534/176030821-46987de3-e0f3-4dab-b53f-6394d1a873ee.gif)

After:
![test](https://user-images.githubusercontent.com/38889534/176031049-ba1c56bb-bdc9-465b-86f0-a61c5b508dca.gif)


### TESTING INSTRUCTIONS
1. In SQL Editor, enter valid query
2. Result tab should show
3. Click on Explore in Result tab
4. Select overwrite existing and save & explore
5. User should be able to see a list of virtual datasets to choose from to overwrite without typing

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [X] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

Antonio-RiveroMartnez committed:
Save Dataset Modal:

- Use our Select component as substitute of the Autocomplete one so options are loaded initially without the user having to trigger a search and we are mosre consistent with the rest of the app
- Changing datasetId to lowercase so when custom props get into the DOM we don't get warning related to invalid formatting
- We extracted the dropdown out of the radio because it causes invalid click handling when an option is selected
- Updated tests

michael-s-molina commented on superset-frontend/src/SqlLab/components/SaveDatasetModal/index.tsx:
```diff
@@ -338,23 +343,27 @@ export const SaveDatasetModal: FunctionComponent<SaveDatasetModalProps> = ({
+                    filterOption={filterAutocompleteOption}
+                    disabled={newOrOverwrite !== 2}
+                    // @ts-ignore
+                    getPopupContainer={() => document.body}
```
`document.body` is the default value for `getPopupContainer` so you don't need to set it here 😉 

<img width="1240" alt="Screen Shot 2022-06-27 at 6 01 45 PM" src="https://user-images.githubusercontent.com/70410625/176034952-cfbd90eb-30a1-4098-a5ff-c622597d2960.png">

michael-s-molina replied:
Oh, forget it. I just saw that you're overriding the default behavior of the component 😉 You can leave as it is 👍🏼 I need some coffee

Antonio-RiveroMartnez replied:
If I remove it then `Select` component will use its current behavior (parent node): https://github.com/apache/superset/blob/master/superset-frontend/src/components/Select/Select.tsx#L704

michael-s-molina commented on superset-frontend/src/SqlLab/components/SaveDatasetModal/index.tsx:
```diff
@@ -158,49 +162,56 @@ export const SaveDatasetModal: FunctionComponent<SaveDatasetModalProps> = ({
+  const loadDatasetOverwriteOptions = useCallback(
+    async (input = '') => {
+      const { userId } = user;
+      if (userId) {
```
Looking at the `User` type, there's a comment that guest users don't have an `userId`. My question is, can guest users access this part of the application? It would be weird that a guest user is able to save a dataset. If they can't then we can assume that the `userId` always exists and remove this `if` statement.

Antonio-RiveroMartnez replied:
Oh great! I was just respecting as much logic from before when hitting the API. But let me confirm that and remove such `if` if that's the case. Thanks

Antonio-RiveroMartnez committed:
Save Dataset Modal:

- Update missing test for DatasourceControl

Antonio-RiveroMartnez committed:
Save Dataset Modal:

- Remove conditional from load options function since only guest users dont have userId, and if that is the case they wont reach this part of the application

stephenLYZ reviewed (commented):
Thanks for the fix! Left some comments

Antonio-RiveroMartnez commented on superset-frontend/src/SqlLab/components/SaveDatasetModal/index.tsx:
```diff
@@ -139,7 +143,7 @@ export const SaveDatasetModal: FunctionComponent<SaveDatasetModalProps> = ({
   const handleOverwriteDataset = async () => {
     await updateDataset(
       query.dbId,
-      datasetToOverwrite.datasetId,
```
This custom prop gets into the DOM so react throws a warning for invalid format. That's why it was changed to all lowercase.

Antonio-RiveroMartnez commented on superset-frontend/src/SqlLab/components/SaveDatasetModal/index.tsx:
```diff
@@ -338,23 +334,27 @@ export const SaveDatasetModal: FunctionComponent<SaveDatasetModalProps> = ({
+                    value={selectedDatasetToOverwrite}
+                    filterOption={filterAutocompleteOption}
+                    disabled={newOrOverwrite !== 2}
+                    // @ts-ignore
```
`getPopupContainer` is not a prop in `Select` component because it already has its own, however, using such default behavior on a modal with not enough height causes the option to render above the input. So we need to override such behavior for this particular case and thus send it as a prop.

Antonio-RiveroMartnez replied:
Yes, I'd assume we can and it as an optional prop. Will get back to you guys when it's updated

Antonio-RiveroMartnez committed:
Save Dataset Modal:

- Remove unused comment

Antonio-RiveroMartnez committed:
Save Dataset Modal:

- Add getPopupContainer as prop for Select component

Antonio-RiveroMartnez committed:
Save Dataset Modal:

- Add tests for our new getPopupContainer prop in Select component. So if passed it gets called.

diegomedina248 commented on superset-frontend/src/components/Select/Select.tsx:
```diff
@@ -165,6 +165,11 @@ export interface SelectProps extends PickedSelectProps {
+   * Customize who the container node is.
+   * Parent node by default.
+   */
+  getPopupContainer?: (triggerNode?: any) => any;
```
I think the return should be `HTMLElement`

Antonio-RiveroMartnez replied:
Yeah, I had it like that, but SelectFilterPlugin is also customizing such prop and I didn't want to impact other non-related component.

rusackas reviewed (approved):
LGTM

rusackas commented:
Just needs a rebase, and I'll happily merge :)

Antonio-RiveroMartnez committed:
Save Dataset Modal:

- use lowercased property when calling post form data

jinghua-qa commented:
/testenv up

michael-s-molina commented on superset-frontend/src/components/Select/Select.tsx:
```diff
@@ -165,6 +165,11 @@ export interface SelectProps extends PickedSelectProps {
+   * Customize who the container node is.
+   * Parent node by default.
+   */
+  getPopupContainer?: (triggerNode?: any) => any;
```
`getPopupContainer` comes from Ant Design. Can you include it in `PickedSelectProps` instead?

Antonio-RiveroMartnez replied:
If I do that, then the definition in `src/filters/components/Select/SelectFilterPlugin.tsx` will be invalid:
`(trigger: HTMLElement) => ParentNode | null' is not assignable to type 'RenderDOMFunc'`

And I didn't want to touch any other non-related component. But if you think I must, I will.

michael-s-molina replied:
I think it's important to preserve Ant Design's original signature. You can change `src/filters/components/Select/SelectFilterPlugin.tsx` to:

```
getPopupContainer={
  showOverflow
    ? () => (parentRef?.current as HTMLElement) || document.body
    : (trigger: HTMLElement) =>
        (trigger?.parentNode as HTMLElement) || document.body
}
```

Antonio-RiveroMartnez replied:
Updated

michael-s-molina commented on superset-frontend/src/components/Select/Select.test.tsx:
```diff
@@ -829,6 +829,28 @@ test('async - requests the options again after clearing the cache', async () =>
 });
 
+test('async - triggers getPopupContainer if passed', async () => {
+  const getPopupContainer = jest.fn(() => null);
```
```suggestion
  const getPopupContainer = jest.fn();
```

Antonio-RiveroMartnez replied:
Updated

michael-s-molina commented on superset-frontend/src/components/Select/Select.test.tsx:
```diff
@@ -829,6 +829,28 @@ test('async - requests the options again after clearing the cache', async () =>
+});
+
+test('static - triggers getPopupContainer if passed', async () => {
+  const getPopupContainer = jest.fn(() => null);
```
```suggestion
  const getPopupContainer = jest.fn();
```

Antonio-RiveroMartnez replied:
Updated

Antonio-RiveroMartnez committed:
Save Dataset Modal:

- Update tests so there is no need to define a null returning func

Antonio-RiveroMartnez committed:
Save Dataset Modal:

- Including getPopupContainer from PickedSelectProps instead
- Updating definition in SelectFilterPlugin
